main.go

```diff
@@ -50,43 +74,8 @@ func main() {
 		r.RequestURI = ""
 		r.URL.Scheme = "http"
 
-		go func() {
-			req_a := *r
-			req_a.URL.Host = ua.Host
-			req_a.URL, _ = url.Parse(r.URL.String())
-			req_a.Cancel = cancel_a
-
-			rt := &http.Transport{DisableKeepAlives: true}
-			resp, err := rt.RoundTrip(&req_a)
-			if err != nil {
-				log.Printf("target A failed: %s", err)
-				close(fail_a)
-			} else if resp.StatusCode >= 500 || resp.StatusCode == 408 {
-				log.Printf("target A unsatisfying status: %d", resp.StatusCode)
-				close(fail_a)
-			} else {
-				resp_a <- resp
-			}
-		}()
-
-		go func() {
-			req_b := *r
-			req_b.URL, _ = url.Parse(r.URL.String())
-			req_b.URL.Host = ub.Host
-			req_b.Cancel = cancel_b
-
-			rt := &http.Transport{DisableKeepAlives: true}
-			resp, err := rt.RoundTrip(&req_b)
-			if err != nil {
-				log.Printf("target B failed: %s", err)
-				close(fail_b)
-			} else if resp.StatusCode >= 500 || resp.StatusCode == 408 {
-				log.Printf("target B unsatisfying status: %d", resp.StatusCode)
-				close(fail_b)
-			} else {
-				resp_b <- resp
-			}
-		}()
+		go forwardRequest(r, ua, resp_a, cancel_a, fail_a)
+		go forwardRequest(r, ub, resp_b, cancel_b, fail_b)
```

Comment:
er, wait. no. `cancel_a` is the channel that cancels request a, so this is correct.